examples/tabular_q.py

```diff
@@ -0,0 +1,169 @@
+flags.DEFINE_integer("episodes", 5000, "number of episodes used to learn.")
+flags.DEFINE_float("epsilon", 0.2, "Epsilon rate of exploration. ")
+flags.DEFINE_integer("episode_length", 5, "The number of steps in each episode.")
+FLAGS = flags.FLAGS
+
+
+def hash_state_action(autophase_feature, action, step):
+    return tuple(
+        (
+            tuple(autophase_feature[FLAGS.features_indices]),
+            step,
+            FLAGS.actions.index(action),
+        )
+    )
+
+
+def get_env():
+    return env_from_flags(benchmark_from_flags())
+
+
+def select_action(q_table, ob, step, epsilon=0.0):
+    qs = [q_table.get(hash_state_action(ob, act, step), -1) for act in FLAGS.actions]
+    if random.random() < epsilon:
+        return random.choice(FLAGS.actions)
+    max_indices = [i for i, x in enumerate(qs) if x == max(qs)]
+    # Breaking ties at random by selecting any of the indices.
+    return FLAGS.actions[random.choice(max_indices)]
+
+
+def get_max_q_value(q_table, ob, step):
+    max_q = 0
+    for act in FLAGS.actions:
+        hashed = hash_state_action(ob, act, step)
+        if hashed not in q_table:
+            q_table[hashed] = 0
+        max_q = max(q_table[hashed], max_q)
+    return max_q
+
+
+def train(q_table, env):
+    # Buffer an old version of q table to inspect training progress
+    prev_q = {}
+
+    for i in range(FLAGS.episodes):
+        current_length = 0
+        obs = env.reset()
+        while current_length < FLAGS.episode_length:
+            # Run Epsilon greedy policy.
+            a = select_action(q_table, obs, current_length, FLAGS.epsilon)
+            hashed = hash_state_action(obs, a, current_length)
+            if hashed not in q_table:
+                q_table[hashed] = 0
+            obs, reward, done, info = env.step(env.action_space.flags.index(a))
+            current_length += 1
+            # Get max q at new state.
+            target = reward + FLAGS.discount * get_max_q_value(
+                q_table, obs, current_length
+            )
+            # Update Q value at current state action pair.
+            q_table[hashed] = (
+                FLAGS.learning_rate * target
+                + (1 - FLAGS.learning_rate) * q_table[hashed]
+            )
+
+        if i % 50 == 0:
+            print(f"Running episode {i}")
+
+            def compare_qs(q_old, q_new):
+                diff = [q_new[k] - v for k, v in q_old.items()]
+                return sum(diff) / len(diff) if diff else "NaN"
+
+            difference = compare_qs(prev_q, q_table)
+            print(
+                f"Newly added Q entries {len(q_table)-len(prev_q)}, averaged diff {difference}"
+            )
+            prev_q = q_table.copy()
+
+
+def setup_env():
+    FLAGS.observation = "Autophase"
+    FLAGS.reward = "IrInstructionCount"
+    FLAGS.benchmark = "cBench-v0/dijkstra"
+    FLAGS.env = "llvm-ic-v0"
+
+
+def main(argv):
+    setup_env()
+    # Train a Q table.
+    q_table = {}
+    env = get_env()
```

Comment:
I would suggest using the `gym.make(...)` interface to set up the environment rather than setting commandline flags at runtime. If we use the environment `"llvm-autophase-ic-v0"` this already has the observation and reward spaces set up how you want. IMO this is easier to read, and also makes it clear to the user which flags can and cannot be overriden (now, only the `--benchmark` can be set):

```suggestion
    q_table: Dict[StateActionTuple, Reward] = {}
    benchmark = benchmark_from_flags()
    assert benchmark, "You must specify a benchmark using the --benchmark flag"
    env = gym.make("llvm-autophase-ic-v0", benchmark=benchmark)
```

I also added type hints to the `q_table` because I couldn't help myself 😃 If you go with this then the `setup_env()` and `get_env()` functions are not required, along with the `env_from_flags` import.